tools/lib/port_cleanup.py

```diff
@@ -0,0 +1,272 @@
+    """
+    # TODO: Need to also handle the KS v3 session approach.
+    # Port it from openstack mojo helpers, then make those use these.
+    novarc = get_novarc()
+    auth = {
+        'username': novarc['OS_USERNAME'],
+        'password': novarc['OS_PASSWORD'],
+        'auth_url': novarc['OS_AUTH_URL'],
+        'project_name': novarc['OS_TENANT_NAME'],
+        'region_name': novarc['OS_REGION_NAME'],
+        'insecure': True,
+        'version': 2,
+    }
+    return auth
+
+
+def get_nova_client():
+    """Get nova client
+    """
+    auth = get_auth()
+    nc = novaclient.Client(**auth)
+    assert check_nova_client(nc) is True
+    return nc
+
+
+def get_keystone_client(verison=None):
+    """Get keystone client
+    """
+    auth = get_auth()
+    kc = keystoneclient.Client(**auth)
+    assert check_ks_client(kc) is True
+    return kc
+
+
+def get_neutron_client():
+    """Get neutron client
+    """
+    auth = get_auth()
+    nc = neutronclient.Client(**auth)
+    assert check_neutron_client(nc) is True
+    return nc
+
+
+def check_ks_client(ks=None):
+    """Check keystone client with a simple query. Also expect a region name.
+    """
+    _check = ks.service_catalog.region_name
+    logging.info('Keystone client region name check: '
+                 '{}'.format(_check is not None))
+    return _check is not None
+
+
+def check_neutron_client(nc=None):
+    """Check neutron client with a simple query for networks. Also expect
+    one or more networks to exist.
+    """
+    _check = nc.list_networks()['networks']
+    logging.info('Neutron client network list check: {}'.format(len(_check)))
+    return len(_check) > 0
+
+
+def check_nova_client(nc=None):
+    """Check nova client with a simple query for flavors. Also expect
+    one or more flavors to exist.
+    """
+    _check = nc.flavors.list()
+    logging.info('Nova client flavor list check: {}'.format(len(_check)))
+    return len(_check) > 0
+
+
+def get_servers_in_state(clients, status=None):
+    """Get a list of nova servers (instances) which are in a specific state.
+    """
+
+    if status is None:
+        status = 'ERROR'
+
+    _servers = []
+    for instance in clients['nv'].servers.list():
+        if instance.status == status:
+            _servers.append(instance)
+    logging.info('Nova servers in {} status: {}'.format(status, len(_servers)))
+    return _servers
+
+
+def get_ports(clients, name_regex=None, status=None):
+    """Get a list of neutron ports, or a subset of ports which match
+    a pattern by name.
+    """
+
+    if not name_regex:
+        name_regex = ""
+    re_name = re.compile(name_regex)
+
+    ports = clients['nu'].list_ports()['ports']
+    logging.info('Neutron ports total: {}'.format(len(ports)))
+
+    _ports_matched = []
+    for port in ports:
+        if re_name.match(port['name']) and not status:
+            _ports_matched.append(port)
+
+        if (re_name.match(port['name']) and status and
```

Comment:
These can be combined with an OR since you are doing the same thing in both places:

if (re_name.match(port['name']) and not status) or (re_name.match(port['name']) and status and port['status'] == status):